### ISSUE TITLE
Question on urls and routes

### ISSUE THREAD

**Issue by onekiloparsec** [hishnash/djangochannelsrestframework#9]
Hi. I just followed the path from `channels_api` to here, and thanks for your efforts! I am using Django2, and wanted to added websocket REST API endpoints to my backend.

But I am a bit lost by all the possibilities. I have a backend with multiple django apps inside. Say, one of these is called `activities` and its urls are accessible under `http://api.example.com/activities/`. 

I would like for instance to subsribe to the creation of new activities. I wondered whether I can use my DRF API views over websocket, as the end of your README seems to suggest (with `view_as_consumer`). ?

Or should I use some other combination, with for instance the HyperMediaChannels ? 

Thanks a lot for your help.

**Comment by hishnash** [comment]
currently, there is nothing out of the box that lets you subscribe to a list (or filtered list) of objects being created. (this is on my todo list)

however, if you want to subscribe to all `activities` or can split these up (without knowing the filter) you could do this like this:

```python
from djangochannelsrestframework.consumers import AsyncAPIConsumer
from djangochannelsrestframework.observer import model_observer

class ActivityConsumer(AsyncAPIConsumer):
    async def accept(self):
        # subscribe to all new changed events
        await ActivityConsumer. activities_change.subscribe(self)
        await super().accept()

    @model_observer(Activitie)  # you Django Model cls goes here.
    async def activities_change(self, message, observer=None, **kwargs):
        # check the message for the action type if you only want to send events for new ones.
        await self.send_json(message)
```

> as the end of your README seems to suggest (with view_as_consumer). ?

they `view_as_consumer` does not support subscribtion to events it is there just to let you expose the same views as you do over classic HTTP but over a websocket. (is most usefull when combined with the multiplexing lib that lets you have one websocket connection rout messages to different consuemrs/views)

**Comment by hishnash** [comment]
> Or should I use some other combination, with for instance the HyperMediaChannels ?

HyperMediaChannels is really just focusing on how to serialise a model instance in such a way that any foreign key references are returned as prototype json messages that could be sent down the WebSocket to retrieve the related object (just as a hypermedia http api would return a URL)

**Comment by onekiloparsec** [comment]
Thanks a lot for your explanations, I start to understand a bit the vocabulary of all this.

I followed your suggestion. Here is my `routing.py`:

```python
class ActivityConsumer(AsyncAPIConsumer):
    async def accept(self, subprotocol=None):
        # subscribe to all new changed events
        await ActivityConsumer.activities_change.subscribe(self)
        await super().accept()

    @model_observer(Activity)  # you Django Model cls goes here.
    async def activities_change(self, message, observer=None, **kwargs):
        # check the message for the action type if you only want to send events for new ones.
        await self.send_json(message)


application = ProtocolTypeRouter({
    "websocket": AuthMiddlewareStack(
        URLRouter([
            path("activities/live/", ActivityConsumer),
        ])
    ),
})
```

Then, when I try to connect with a websocket like that:

```javascript
var ws = new WebSocket("ws://api.lvh.me:8000/activities/live/")
```

it fails with an error:

```
2018-08-19 10:02:12,595 - INFO - server - HTTP/2 support not enabled (install the http2 and tls Twisted extras)
2018-08-19 10:02:12,596 - INFO - server - Configuring endpoint tcp:port=8000:interface=127.0.0.1
2018-08-19 10:02:12,597 - INFO - server - Listening on TCP address 127.0.0.1:8000
[2018/08/19 10:02:41] WebSocket HANDSHAKING /activities/live/ [127.0.0.1:64624]
2018-08-19 10:02:42,656 - ERROR - server - Exception inside application: 'NoneType' object has no attribute 'group_add'
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/sessions.py", line 175, in __call__
    return await self.inner(receive, self.send)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/middleware.py", line 41, in coroutine_call
    await inner_instance(receive, send)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/consumer.py", line 56, in __call__
    await await_many_dispatch([receive], self.dispatch)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/utils.py", line 50, in await_many_dispatch
    await dispatch(result)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/consumer.py", line 67, in dispatch
    await handler(message)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/generic/websocket.py", line 173, in websocket_connect
    await self.connect()
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/channels/generic/websocket.py", line 180, in connect
    await self.accept()
  File "/Users/onekiloparsec/code/arcsecond-back/arcsecond/routing.py", line 15, in accept
    await ActivityConsumer.activities_change.subscribe(self)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/djangochannelsrestframework/observer/observer.py", line 56, in subscribe
    await consumer.add_group(group_name)
  File "/Users/onekiloparsec/.virtualenvs/arcsecond-back-py364/lib/python3.6/site-packages/djangochannelsrestframework/consumers.py", line 70, in add_group
    await self.channel_layer.group_add(
  'NoneType' object has no attribute 'group_add'
[2018/08/19 10:02:42] WebSocket DISCONNECT /activities/live/ [127.0.0.1:64624]
```

It seems somehow `self.channel_layer` is not yet created. This is a bug report now... :-) I'd happy to help once I feel better confortable with all this.

**Comment by onekiloparsec** [comment]
This is live debugging ! Thanks. I was missing this channel layer, indeed. So I added it to my settings.

No more errors, I can connect to my websocket, wonderful.

However, nothing happen anymore. Should I send `subscribe_instance` (tried with and without). I have set this up too:

```javascript
ws.onmessage = function (m) { console.log(m) }
```

to see results.

**Comment by hishnash** [comment]
`subscribe_instance` is for subscribing to changes on a single instance. you should be Subscribed to all model_save/delete events on the  `Activity` model. 

maybe add a print statement in `activities_change` just to check it is called and then with your server running (using maybe Django admin) edit the model. (changes need to come through Django this does not observe the db directly just the pre-save and post-save events fired by Django)

**Comment by onekiloparsec** [comment]
Indeed `activities_change` is not called (I've a breakpoint on it that's never hit).

**Comment by hishnash** [comment]
hmm so you connect then in Django admin (on the same server) create an Activity instance?

**Comment by onekiloparsec** [comment]
Sorry, it is called when I use the Django admin interface, but not when I use a Python console. Which... is logical. Sorry, my mistake.

**Comment by onekiloparsec** [comment]
Wow, that opens up a world of possibilties now... :-) Thanks a lot!

**Comment by hishnash** [comment]
if you need to say just subscribe/limit the subscription a little you can add another method to your example. (say Activity has a field `event_type` and you want to just subscribe to Activiteis of event type you could do this.)

```python
from djangochannelsrestframework.consumers import AsyncAPIConsumer
from djangochannelsrestframework.observer import model_observer, ModelObserver


class ActivityConsumer(AsyncAPIConsumer):
    async def accept(self, subprotocol=None):
        # subscribe to all new changed events
        await ActivityConsumer.activities_change.subscribe(self, event_type='party')
        await super().accept()

    @model_observer(Activity)  # you Django Model cls goes here.
    async def activities_change(self, message, observer=None, **kwargs):
        # check the message for the action type if you only want to send events for new ones.
        await self.send_json(message)

    @activities_change.groups
    def activities_change(self: ModelObserver, instance, consumer=None, event_type=None, *args, **kwargs):
        # this method is called once for each change to work out what groups to use to send the events down.
        # if being called above by the subscribed event.
        if consumer:
            return f'type:{event_type}'
        else:
            # we have an instance and we need to know what group to send the 
            # message down.
            yield f'type:{instance.event_type}'
```

**Comment by onekiloparsec** [comment]
Hm, interesting. There is 2 different things in your new example. The `event_type` stuff, but also the fact that we have or not an instance. 

I am trying to access that instance of `Activity` inside `activities_change` (to send the serialized activity down to the socket) but it seems I don't always have it (although I get a `pk`). Is that because we get `pre_save` or `post_save` ?

**Comment by hishnash** [comment]
> on pre_save or post_save ?

you should only not be able to access it on delete events. or if you are saving as part of a transaction it is on my todo list to have a solution that matches up events produced mid transaction and send them only once the transaction commits.

**Comment by hishnash** [comment]
> There is 2 different things in your new example. The event_type stuff, but also the fact that we have or not an instance.

so the group's method is called once for pre-save and once for post-save with the instance. (this means that if due to a change a given instance is no longer in a group but know in a new group a delete event is sent to the first group and a created event to the second). 

but we also call the groups method when you subscribe (and pass through any kwargs passed to the subscribe method). this binds the consumer to the groups yielded at this point.

**Comment by hishnash** [comment]
> to send the serialized activity down to the socket

what you can do is use the serializer binding to do this work before sending the object over the group so that you have the data even though it is stuck within a transaction.

```
@activities_change.serializer
def activities_change(self: ModelObserver, instance, action, **kwargs):
    # return a dict 
```
